### PR TITLE
Driver error injection and handling

### PR DESCRIPTION
**Pull request by avrahamshukron**
## We now handle all kinds of errors in case a sensor fails.

image

## Usage
If you want to simulate errors, pass a value between 0-1.0 to `--error` command line option.
A good value is 0.001 (error every 1000 reads)

### CODE REVIEW

File: drivers/mocks/sensor.py [hunk 1]
```diff
@@ -7,8 +8,18 @@ class MockSensor(object):
     sequence. The sequence is replayed indefinitely.
     """
 
-    def __init__(self, seq):
+    def __init__(self, seq, error_probability=0):
         self.data = cycle(seq)
+        self.error_probability = error_probability
+
+    def random_error(self):
+        exe = random.choice([ValueError, OSError, TimeoutError, ZeroDivisionError])
```
Comment (rinslow) [inline drivers/mocks/sensor.py:16]: lol creative

File: main.py [hunk 2]
```diff
@@ -96,7 +100,8 @@ def main():
         log.info("Running in simulation mode! simulating: "
                  "flow, pressure sensors, and watchdog")
```
Comment (rinslow) [inline main.py:101]: add error probablility in log

File: algo.py [hunk 3]
```diff
@@ -277,6 +277,46 @@ def reset_min_values(self):
...
+            error = True
+            self._events.alerts_queue.enqueue_alert(AlertCodes.PRESSURE_SENSOR_ERROR)
+            self.log.error(e)
+
+        try:
+            o2_saturation_percentage = self._oxygen_a2d.read()
+        except Exception as e:
+            error = True
+            self._events.alerts_queue.enqueue_alert(AlertCodes.SATURATION_SENSOR_ERROR)
+            self.log.error(e)
```
Comment (rinslow) [inline algo.py:313]: kind of ugly and repetitive, 
extract that to a method that happens for each sensor

**Comment by rinslow** [comment]
Very good job overall, but if we are into simulating things, most of the errors happen when the sensors fail to even initialize themselves, how about a flag simulating this as well? could be useful

File: drivers/mocks/sensor.py (hunk 1 above)
Comment (avrahamshukron) [reply drivers/mocks/sensor.py:16]: ;)

File: main.py (hunk 2 above)
Comment (avrahamshukron) [reply main.py]: done

**Comment by avrahamshukron** [comment]
> Very good job overall, but if we are into simulating things, most of the errors happen when the sensors fail to even initialize themselves, how about a flag simulating this as well? could be useful

I think it should be done in another issue, as it is another feature.

File: algo.py (hunk 3 above)
Comment (avrahamshukron) [reply algo.py:313]: Absolutely right. Will fix

Comment (avrahamshukron) [reply algo.py]: done

**Pull request merged**